src/adapter/src/coord.rs

```diff
@@ -867,6 +1011,24 @@ pub struct Coordinator {
 
     /// Tracing handle.
     tracing_handle: TracingHandle,
+
+    /// Information about statement executions that have been logged
+    /// but not finished.
+    ///
+    /// This map needs to have enough state left over to later retract
+    /// the system table entries (so that we can update them when the
+    /// execution finished.)
+    statement_logging_executions_begun: BTreeMap<Uuid, StatementBeganExecutionRecord>,
+
+    /// Information about sessions that have been started, but which
+    /// have not yet been logged in `mz_session_history`.
+    /// They may be logged as part of a statement being executed (and chosen for logging).
+    statement_logging_unlogged_sessions: BTreeMap<Uuid, SessionHistoryEvent>,
```

Comment:
It's a bit hard to imagine these growing too huge without actively malicious behavior (E.g., executing billions of statements that never terminate).

The `statement_logging_execution_begun` entries are removed whenever the statement finishes (things could pile up if there's a bug where we forgot to retire some statement, but the `soft_assert_or_log` in `ExecuteContextExtra` is meant to make that harder). The `statement_logging_unlogged_sessions` stick around until the next reboot (or until the session issues a statement that is logged), but given that by far our biggest user of sessions uses at most a few thousand per day it will only grow to ~15000 max between maintenance windows.

One thing I can do make the risk even less is to drop sessions from this map when they disconnect, because there's no good reason not to.